align/pnr/checker.py

```diff
@@ -54,3 +59,87 @@ def check_placement(placement_verilog_d, scale_factor):
                         ury=bbox.ury/scale_factor
                     )
                 )
+
+
+def _transform_leaf(module, instance, leaf):
+    if 'transformation' in leaf:
+        tr_leaf = transformation.Transformation(**leaf['transformation'])
+    else:
+        tr_leaf = transformation.Transformation(**{'oX': leaf['bbox'][0], 'oY': leaf['bbox'][1], 'sX': 1, 'sY': 1})
+    tr_inst = transformation.Transformation(**instance['transformation'])
+    if 'name' in leaf:
+        name = leaf['name']
+    else:
+        name = leaf['concrete_name']
+    flat_leaf = dict()
+    flat_leaf['concrete_name'] = leaf['concrete_name']
+    flat_leaf['name'] = instance['instance_name'] + '/' + name
+    flat_leaf['transformation'] = transformation.Transformation.mult(tr_inst, tr_leaf).toDict()
+    module['leaves'].append(flat_leaf)
+
+
+def _flatten_leaves(placement, concrete_name):
+    """ transform leaf coordinates to top level """
+    module = placement['modules'][concrete_name]
+    module['leaves'] = []
+    for instance in module['instances']:
+        leaf = placement['leaves'].get(instance['concrete_template_name'], False)
+        if leaf:
+            _transform_leaf(module, instance, leaf)
+        else:
+            leaves = _flatten_leaves(placement, instance['concrete_template_name'])
+            for leaf in leaves:
+                _transform_leaf(module, instance, leaf)
+    return module['leaves']
+
+
+def _check_place_on_grid(leaf, constraints):
```

Comment:
@854768750 , this might be of interest to you. This PR is baking this check into the flow.